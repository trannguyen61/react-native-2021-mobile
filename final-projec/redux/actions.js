import { fetchMovies } from '../api/movies';
import { login } from '../api/auth'

export const SEARCH_QUERY_SENT = 'SEARCH_QUERY_SENT';
export const NEW_RESULTS_RECEIVED = 'NEW_RESULTS_RECEIVED';
export const ADDITIONAL_RESULTS_RECEIVED = 'ADDITIONAL_RESULTS_RECEIVED';
export const RESULTS_ERROR = 'RESULTS_ERROR';
export const LOGIN = 'LOGIN'
export const LOGIN_FAILED = 'LOGIN_FAILED'
export const LOGOUT = 'LOGOUT'

export const onLogin = (username, password, loginFn = login) => async (dispatch) => {
  const response = loginFn(username, password)
  if (response.token) {
    dispatch({
      type: LOGIN,
      payload: { token: response.token }
    })
  } else {
    dispatch({
      type: LOGIN_FAILED,
      payload: { loginErr: response.err }
    })
  }
}

export const onFetchMovies = (search, pageNumber, fetchFn = fetchMovies) => async (dispatch) => {
  dispatch({ type: SEARCH_QUERY_SENT });
  try {
    const response = await fetchFn(search, pageNumber);
    const results = response.results;

    const totalResults = response.totalResults;
    const maxPages = Math.ceil(totalResults / 10); // 10 results per page

    if (pageNumber === 1) {
      dispatch({
        type: NEW_RESULTS_RECEIVED,
        payload: { movies: results, maxPages: maxPages },
      });
    } else {
      dispatch({
        type: ADDITIONAL_RESULTS_RECEIVED,
        payload: { movies: results },
      });
    }
  } catch (err) {
    dispatch({ type: RESULTS_ERROR, payload: {
        err: 'There\'s something wrong with the server. Please try again.'
    } });
  }
};