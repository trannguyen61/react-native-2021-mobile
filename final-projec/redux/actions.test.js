import * as actions from './actions';

describe('login returns response', () => {
    const mockFetch = (username, password) => {
        if (username == 'a' && password == 'a') {
          return { token: 'token' }
        }
    
        return { err: 'Wrong credentials.' }
    };

    it('dispatches LOGIN when successfully logged in', async () => {
        const mockDispatch = jest.fn();
        await actions.onLogin('a', 'a', mockFetch)(mockDispatch);
        expect(mockDispatch.mock.calls[0][0]).toEqual({
          type: actions.LOGIN,
          payload: { token: 'token' },
        });
        expect(mockDispatch.mock.calls[1]).toMatchSnapshot();
    });

    it('dispatches LOGIN_FAILED with wrong credentials', async () => {
        const mockDispatch = jest.fn();
        await actions.onLogin('a', 'abc', mockFetch)(mockDispatch);
        expect(mockDispatch.mock.calls[0][0]).toEqual({
          type: actions.LOGIN_FAILED,
          payload: { loginErr: 'Wrong credentials.' },
        });
        expect(mockDispatch.mock.calls[1]).toMatchSnapshot();
    });
})

describe('onFetchMovies returns actions', () => {
  const mockFetch = (searchTerm, pageNumber) => {
    if (searchTerm === 'test' && (pageNumber === 1 || pageNumber == 2)) {
      return { results: [], totalResults: 20 };
    }

    throw new Error(errMessage);
  };

  it('dispatches SEARCH_QUERY_SENT', async () => {
    const mockDispatch = jest.fn();
    await actions.onFetchMovies('', '')(mockDispatch);
    expect(mockDispatch.mock.calls[0][0]).toEqual({
      type: actions.SEARCH_QUERY_SENT,
    });
  });

  it('dispatches NEW_RESULTS_RECEIVED when has search query and page 1', async () => {
    const mockDispatch = jest.fn();
    await actions.onFetchMovies('test', 1, mockFetch)(mockDispatch);
    expect(mockDispatch.mock.calls[1][0]).toEqual({
      type: actions.NEW_RESULTS_RECEIVED,
      payload: { movies: [], maxPages: 2 },
    });
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot();
  });

  it('dispatches ADDITIONAL_RESULTS_RECEIVED when has search query and page 2', async () => {
    const mockDispatch = jest.fn();
    await actions.onFetchMovies('test', 2, mockFetch)(mockDispatch);

    expect(mockDispatch.mock.calls[1][0]).toEqual({
      type: actions.ADDITIONAL_RESULTS_RECEIVED,
      payload: { movies: [] },
    });
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot();
  });

  it('dispatches RESULTS_ERROR with incorrect arguments', async () => {
    const mockDispatch = jest.fn();
    await actions.onFetchMovies(test, 1000, mockFetch)(mockDispatch);

    expect(mockDispatch.mock.calls[1][0]).toEqual({
      type: actions.RESULTS_ERROR,
      payload: { err: "There's something wrong with the server. Please try again." },
    });
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot();
  });
});