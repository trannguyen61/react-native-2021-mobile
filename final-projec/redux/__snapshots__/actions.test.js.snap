// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login returns response dispatches LOGIN when successfully logged in 1`] = `undefined`;

exports[`login returns response dispatches LOGIN_FAILED with wrong credentials 1`] = `undefined`;

exports[`onFetchMovies returns actions dispatches ADDITIONAL_RESULTS_RECEIVED when has search query and page 2 1`] = `
Array [
  Object {
    "payload": Object {
      "movies": Array [],
    },
    "type": "ADDITIONAL_RESULTS_RECEIVED",
  },
]
`;

exports[`onFetchMovies returns actions dispatches ADDITIONAL_RESULTS_RECEIVED with search term and page 2 1`] = `
Array [
  Object {
    "payload": Object {
      "movies": Array [],
    },
    "type": "ADDITIONAL_RESULTS_RECEIVED",
  },
]
`;

exports[`onFetchMovies returns actions dispatches NEW_RESULTS_RECEIVED when has search query and page 1 1`] = `
Array [
  Object {
    "payload": Object {
      "maxPages": 2,
      "movies": Array [],
    },
    "type": "NEW_RESULTS_RECEIVED",
  },
]
`;

exports[`onFetchMovies returns actions dispatches RESULTS_ERROR with incorrect arguments 1`] = `
Array [
  Object {
    "payload": Object {
      "err": "There's something wrong with the server. Please try again.",
    },
    "type": "RESULTS_ERROR",
  },
]
`;
